### PR TITLE
Implement a CFG

### LINKED ISSUE #368: feat(semantic): Control Flow Graph

### PR DESCRIPTION
**Pull request by u9g**
close: #368

### CODE REVIEW

File: crates/oxc_linter/src/options.rs [hunk 1]
```diff
@@ -195,7 +195,7 @@ impl LintOptions {
         let mut rules = rules.into_iter().collect::<Vec<_>>();
         // for stable diagnostics output ordering
         rules.sort_unstable_by_key(RuleEnum::name);
-        Ok((rules, LintSettings::default()))
+        Ok((vec![], LintSettings::default()))
```
Comment (Boshen) [inline crates/oxc_linter/src/options.rs:198]: revert?

File: crates/oxc_ast/src/visit.rs [hunk 2]
```diff
@@ -655,6 +655,7 @@ pub trait Visit<'a>: Sized {
     fn visit_elision(&mut self, span: Span) {
         let kind = AstKind::Elision(span);
         self.enter_node(kind);
+        /* cfg - currently elisions are not represented in the cfg */
```
Comment (Boshen) [inline crates/oxc_ast/src/visit.rs:658]: Better document this else where, this `Visit` trait is public facing.

File: crates/oxc_semantic/src/builder.rs [hunk 3]
```diff
@@ -394,6 +413,1902 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
...
+    fn visit_ts_call_signature_declaration(&mut self, signature: &TSCallSignatureDeclaration<'a>) {
+        self.visit_formal_parameters(&signature.params);
+        if let Some(parameters) = &signature.type_parameters {
+            self.visit_ts_type_parameter_declaration(parameters);
+        }
+
+        if let Some(annotation) = &signature.return_type {
+            self.visit_ts_type_annotation(annotation);
+        }
+    }
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:2311]: Can you remove all the visitors that don't have a cfg override right now? Or are you changing the visit order in all these visits?

File: crates/oxc_linter/src/rules/eslint/no_this_before_super.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+#[derive(Debug, Default, Clone)]
+pub struct NoThisBeforeSuper;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Requires all getters to have a return statement
+    ///
+    /// ### Why is this bad?
+    /// Getters should always return a value. If they don't, it's probably a mistake.
```
Comment (u9g) [inline crates/oxc_linter/src/rules/eslint/no_this_before_super.rs:32]: What's a good message to put here and above and below?

File: crates/oxc_semantic/src/builder.rs (hunk 3 above)
Comment (u9g) [reply crates/oxc_semantic/src/builder.rs]: Done.

File: crates/oxc_ast/src/visit.rs (hunk 2 above)
Comment (u9g) [reply crates/oxc_ast/src/visit.rs]: Done.

File: crates/oxc_semantic/src/control_flow.rs
Comment (Boshen) [inline crates/oxc_semantic/src/control_flow.rs:1]: We need all the references and learning materials documented at the top of the file.

**Comment by Boshen** [comment]
I'll merge this when I make a new oxlint release, probably this weekend.

**Comment by Boshen** [comment]
Merged with https://github.com/oxc-project/oxc/pull/2019

This branch will be kept for references.

**Pull request closed without merging**